### PR TITLE
pci: vfio: Implement support for sparse mmap

### PR DESCRIPTION
**Pull request by sboeuf**
Reorganizing the code to leverage the same mechanics implemented for
vfio-user and aimed at supporting sparse memory mappings for a single
region.

Relying on the capabilities returned by the vfio-ioctls crate, we create
a list of sparse areas depending if we get SPARSE_MMAP or MSIX_MAPPABLE
capability, or a single sparse area in case we couldn't find any
capability.

The list of sparse areas is then used to create both the memory mappings
in the Cloud Hypervisor address space and the hypervisor user memory
regions.

This allowed for the simplification of the MmioRegion structure.

### CODE REVIEW

**Comment by sboeuf** [comment]
Depends on https://github.com/cloud-hypervisor/cloud-hypervisor/pull/4023

**Comment by sboeuf** [comment]
I've moved it to draft, just the time to let the baremetal VFIO CI to run.

**Comment by sboeuf** [comment]
VFIO CI ran just fine, this is ready for review.

**Review by rbradford — APPROVED** [review]

File: pci/src/vfio.rs [hunk 1]
```diff
@@ -1173,42 +1190,46 @@ impl VfioPciDevice {
...
+                    };
+
+                    if host_addr == libc::MAP_FAILED {
+                        error!(
+                            "Could not mmap sparse area (offset = 0x{:x}, size = 0x{:x}): {}",
+                            area.offset,
+                            area.size,
+                            std::io::Error::last_os_error()
+                        );
+                        return Err(VfioPciError::MmapArea);
```
Comment (likebreath) [inline pci/src/vfio.rs:1223]: Before return the error, do we need to cleanup the successful mmaps that are already done from the previous iterations of the for loop?

Comment (sboeuf) [reply pci/src/vfio.rs:1223]: I don't think so. The error will cause the vmm thread to terminate, meaning the Cloud Hypervisor process will exit. Therefore all the mmaps will be cleaned by the OS.

Comment (likebreath) [reply pci/src/vfio.rs:1223]: What about the case for hotplug with `add-device`? IIUC, the error would end up with a `HttpError::ApiError` responded to the API requester, while the vmm thread keeps running, right?

Comment (sboeuf) [reply pci/src/vfio.rs:1223]: This is a good point. Could you please open an issue for that? I can take care of it tomorrow.
Do we need to handle the same thing for vfio-user by the way?

Comment (likebreath) [reply pci/src/vfio.rs:1223]: Good point. I thought 'vfio_user' was fine, as it did not return on mmap errors, but it did return on other errors prematurely and will be prone to a similar leak. I just opened an issue #4068 and a quick fix #4069. PTAL.

**Pull request merged**